tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c

```diff
@@ -0,0 +1,331 @@
+};
+
+
+typedef int (*key_value_checker_cb)(const char *key, size_t keylen,
+				    const uint8_t *data, ssize_t datalen);
+
+static bool key_checked;
+
+static int set_load_direct_cb(const char *key,
+			      size_t len,
+			      settings_read_cb read_cb,
+			      void *cb_arg,
+			      void *param)
+{
+	uint8_t data[100];
+
+	LOG_INF("key: %s", key);
+
+	ssize_t bytes = read_cb(cb_arg, &data, sizeof(data));
+
+	LOG_HEXDUMP_INF(data, bytes, "data:");
+
+	struct key_checker_data *validate = (struct key_checker_data *)param;
+
+	if (strcmp(key, validate->key) == 0) {
+		key_checked = true;
+
+		if (!(bytes == validate->exp_value_len &&
+		      memcmp(data, validate->exp_value, bytes) == 0)) {
+			FAIL("Validation failed for key %s", key);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int persistent_values_check(struct key_checker_data *validater)
+{
+
+	key_checked = false;
+	settings_load_subtree_direct("bt", set_load_direct_cb, validater);
+	if (!key_checked) {
+		FAIL("Not found. Key %s", validater->key);
+		return -1;
+	}
+
+	return 0;
+}
+
+/* these data structures are taken from various relavent modules to be checked */
+static struct net_key_val {
+	uint8_t kr_flag : 1,
+		kr_phase : 7;
+	uint8_t val[2][16];
+} __packed;
+
+static struct iv_val {
+	uint32_t iv_index;
+	uint8_t iv_update : 1,
+		iv_duration : 7;
+} __packed;
+
+static struct net_val {
+	uint16_t primary_addr;
+	uint8_t dev_key[16];
+} __packed;
+
+static void test_provisioning_sv_check_save(void)
+{
+	struct key_checker_data validater = { 0 };
+
+	mount_settings_area();
+	bt_mesh_test_cfg_set(&save_cfg, WAIT_TIME);
+	bt_mesh_test_setup(false);
+
+	/* print raw dump of settings file */
+	char dir_path[100] = CONFIG_SETTINGS_FS_DIR;
+
+	read_dir(dir_path);
+
+	/* Verify: check stored provisioning data */
+	/* check stored netkey matches with expected value */
+	struct net_key_val keyval = { 0 };
+
+	keyval.kr_flag = 0;
+	keyval.kr_phase = 0;
+	memcpy(keyval.val[0], save_cfg.net_key, sizeof(save_cfg.net_key));
+
+	snprintk(validater.key, sizeof(validater.key), "mesh/NetKey/%x",
+		 save_cfg.net_idx);
+	validater.exp_value = &keyval;
+	validater.exp_value_len = sizeof(keyval);
+	persistent_values_check(&validater);
+
+	/* check stored addr and devkey matches with expected value */
+	struct net_val netval = { 0 };
+
+	netval.primary_addr = save_cfg.addr;
+	memcpy(netval.dev_key, save_cfg.dev_key, sizeof(save_cfg.dev_key));
+
+	snprintk(validater.key, sizeof(validater.key), "mesh/Net");
+	validater.exp_value = &netval;
+	validater.exp_value_len = sizeof(netval);
+	persistent_values_check(&validater);
+
+	/* check stored IV index params match with expected value */
+	struct iv_val iv;
+
+	iv.iv_index = 0;
+	iv.iv_update = 0;
+	iv.iv_duration = 96;    /* this is as per the mesh stack logic */
+
+	snprintk(validater.key, sizeof(validater.key), "mesh/IV");
+	validater.exp_value = &iv;
+	validater.exp_value_len = sizeof(iv);
+	persistent_values_check(&validater);
```

Comment:
FS Backend is basically 96KB buffer in RAM probably allocated on the executable's stack. Per my understanding, this posix based emulation does not mimic the board's memory layout, so there are no memory constraints as such for the test bench.

I'm not sure what do you mean by proper store/load tests. Store-load emulation is not possible because native posix does not support reboot (soft or hard). Trying to reboot the device terminates emulation. It was found that the only way to test the persistence is to check load and save paths separately as demonstrated here. Agree that this creates a strong coupling with the stack, but it seems like this is the only way. 

I thought about using custom settings backend, however, realized that it does not add any benefit. The test bench will still have to use "settings_save_one" and such calls to save/load data as per the framework of the mesh settings and then validate the data within the testbench using packed structures. So, it does not have much value other than someone can open and see what is in the file (only keys are plain text, the value will still be some kind of packed hex string). A similar thing is done in the current approach by using the "read_dir" function to read files for test-development purpose.

Reply:
Reworked the whole test with a new custom backend. Now the settings are stored in a file on system and this file can be loaded across individual tests.